packages/flutter/lib/src/widgets/widget_span.dart

```diff
@@ -0,0 +1,262 @@
+  @override
+  void build(ui.ParagraphBuilder builder, { double textScaleFactor = 1.0, List<PlaceholderDimensions> dimensions }) {
+    assert(debugAssertIsValid());
+
+    final bool hasStyle = style != null;
+    if (hasStyle)
+      builder.pushStyle(style.getTextStyle(textScaleFactor: textScaleFactor));
+    if (dimensions != null && builder.placeholderCount < dimensions.length) {
+      PlaceholderDimensions currentDimensions = dimensions[builder.placeholderCount];
+      builder.addPlaceholder(
+        currentDimensions.size.width,
+        currentDimensions.size.height,
+        alignment,
+        baseline: currentDimensions.baseline,
+        baselineOffset: currentDimensions.baselineOffset,
+      );
+    }
+    if (hasStyle)
+      builder.pop();
+  }
+
+  /// Calls visitor on this [WidgetSpan]. There are no children spans to walk.
+  bool visitChildren(bool visitor(InlineSpan span)) {
+    if (!visitor(this))
+      return false;
+    if (children != null) {
+      for (InlineSpan child in children) {
+        if (!child.visitChildren(visitor))
+          return false;
+      }
+    }
+    return true;
+  }
+
+  String toPlainText() {
+    return widget.toString();
+  }
+
+  int codeUnitAt(int index) {
+    return null;
+  }
+
+  /// Describe the difference between this widget span and another [TextSpan],
+  /// in terms of how much damage it will make to the rendering. The comparison
+  /// is deep.
+  ///
+  /// Comparing a [WidgetSpan] with a [TextSpan] will result in [RenderComparison.layout].
+  ///
+  /// See also:
+  ///
+  ///  * [TextStyle.compareTo], which does the same thing for [TextStyle]s.
+  RenderComparison compareTo(InlineSpan other) {
+    if (identical(this, other))
+      return RenderComparison.identical;
+    if (other.runtimeType is! WidgetSpan)
+      return RenderComparison.layout;
+    final WidgetSpan typedOther = other;
+    if (typedOther.widget != widget)
+      return RenderComparison.layout;
+    if ((style == null) != (other.style == null))
+      return RenderComparison.layout;
+    RenderComparison result = RenderComparison.identical;
+    if (style != null) {
+      final RenderComparison candidate = style.compareTo(other.style);
+      if (candidate.index > result.index)
+        result = candidate;
+      if (result == RenderComparison.layout)
+        return result;
+    }
+    // WidgetSpans are always a leaf node in the TextSpan tree, and have
+    // no children.
+    return result;
+  }
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other.runtimeType != runtimeType)
+      return false;
+    final WidgetSpan typedOther = other;
+    return typedOther.widget == widget
+        && typedOther.style == style;
+  }
+
+  /// Returns the text span that contains the given position in the text.
+  TextSpan getSpanForPosition(TextPosition position) {
+    assert(debugAssertIsValid());
+    return null;
```

Comment:
shouldn't this be able to return `this` sometimes? What if the `TextPosition` points at the placeholder?

Reply:
I guess the argument is that a TextPosition is always _between_ two characters, and so it can never be _inside_ the PlaceholderSpan, but would always be in its parent?

That makes sense to me.